Fix K2 for new GitHub UI

mjasikowski opened the pull request.

mjasikowski commented:
Made this version 1.4.0 (minor version bump), as it's quite a big change.

tgolen commented on src/js/lib/pages/github/issue.js:
```diff
@@ -13,12 +13,12 @@ import * as API from '../../api';
-    $('.gh-header-actions .k2-element').remove();
-    $('.gh-header-actions').append('<span class="alert k2-element">OOPS!</span>');
+    $('div[data-component="PH_Actions"] .k2-element').remove();
+    $('.div[data-component="PH_Actions"]').append('<span class="alert k2-element">OOPS!</span>');
```
Hm, this doesn't look quite right. The line above has `div` and this line uses `.div`, which would select two very different things, wouldn't it?

mjasikowski replied:
Sorry, that shouldn't have the dot. Fixed!

tgolen commented on src/js/lib/pages/github/issue.js:
```diff
@@ -13,12 +13,12 @@ import * as API from '../../api';
     clearTimeout(clearErrorTimeoutID);
     clearErrorTimeoutID = setTimeout(() => {
-        $('.gh-header-actions .k2-element').remove();
+        $('.div[data-component="PH_Actions"] .k2-element').remove();
```
`.div`?

mjasikowski replied:
Removed the dot

tgolen commented on src/js/lib/pages/github/pr.js:
```diff
@@ -55,54 +55,106 @@ const renderCopyChecklistButton = () => {
+        $('div[data-testid="mergebox-partial"] > div > div > section:last-of-type h3')
+            .text('You cannot merge your own PR');
+        $('div[data-testid="mergebox-partial"] > div > div > section:last-of-type p')
+            .html('It\'s like giving yourself a high-five in public.');
```
🤩

tgolen reviewed (approved):
I'm gonna install this and test it out today!

tgolen commented:
It looks like the issue owner feature isn't working quite right.

![image](https://github.com/user-attachments/assets/a25fa885-a901-4ab7-be3a-5c4e89059236)

## Bug 1
Steps taken:
1. Click the star next to my name
2. Verify that the `Issue Owner` data is added to the GH description

#### Expected outcome:
The star next to my name should now be yellow to indicate that I am the issue owner

#### Actual outcome: 
All the stars remain uncolored

tgolen commented:
## Bug 2
Steps Taken:
1. With one person assigned as the issue owner (as verified by the `Issue Owner` markdown in the GH description)
2. Click the star next to a different assignee to change them to be the current owner

#### Expected Outcome
The old owner should be removed and the new owner should be added.

#### Actual Outcome
The markdown in the GH description contains duplicate `Issue Owner` data for both people.

mjasikowski commented:
Hm, that's weird, I tested the stars earlier today and they were working fine. I'll take another look and see if I can fix that! Thanks for reporting.

MonilBhavsar commented on src/js/lib/pages/github/_base.js:
```diff
@@ -47,11 +45,11 @@ export default function () {
 
     Page.getRepoOwner = function () {
-        return $('.author a span').text();
+        return document.querySelectorAll('.AppHeader-context-item-label.Truncate-text')[0].textContent.trim();
```
NAB, I wonder if we could keep the repeatedly used selectors as constant with meaningful names of what we are selecting - For example - `const selectorForRepoLabel = '.AppHeader-context-item-label.Truncate-text';` , so next time github changes UI, we could simply change the constant instead of going all over the code, and hopefully it should be simple to transition and works 🙏

mjasikowski replied:
For sure, I agree with that. Once we get this PR of the gate I'll spend some time to reorganize the selectors a little bit.

MonilBhavsar replied:
That would be great. Thank you!!

mjasikowski replied:
The repeating selectors are now removed, but I've added a bunch of comments for existing selectors to clarify what exactly they are selecting.

I chose comments over variables to not have to introduce additional null checks to satisfy ESLint for selectors that turn out empty.

MonilBhavsar reviewed (approved):
Thanks for fixing this!

mjasikowski commented:
A few new changes:
- Fixed bugs reported by @tgolen 
- I've modified the owner setting behavior to work based on events with async/await, rather than issue body parsing (for some reason this wasn't working too well)
- Removed the previously-borked-but-fixed behavior that prevented PR self-merging [per this discussion](https://expensify.slack.com/archives/C03TQ48KC/p1736888339349019)
- Added some simple spinner animations for changing the issue owner and clicking checklist copy buttons
- Optimized/simplified the code for setting the owner and rendering checklist copy buttons
- Removed createpr.js, which was unused and contained references to functions that don't exist
- Removed some obsolete rules from .eslintrc.js:

```js
{
        '@lwc/lwc/no-async-await': 'off',
        'es/no-nullish-coalescing-operators' : 'off'
}
```

mjasikowski commented:
Small bug fix, but ready for review now.

tgolen commented on src/js/lib/pages/github/_base.js:
```diff
@@ -7,6 +8,54 @@ import $ from 'jquery';
+        const originalContent = button.innerHTML;
+
+        // Replace the button content with a loader
+        button.innerHTML = '<div class="loader"></div>';
```
```suggestion
        button.innerHTML = '<div class="loader" />';
```

tgolen commented on src/js/lib/pages/github/issue.js:
```diff
@@ -13,118 +13,73 @@ import * as API from '../../api';
-            if (!fileContents) {
-                console.error(`could not load contents of ${pathToChecklist} for some reason`);
-                return;
+async function setOwner(newOwner, oldOwner) {
```
These parameters are confusing. Even after reading the JSDoc, and seeing how the method is implemented, I'm not sure I get it.

I really prefer the simplicity and clarity of the previous methods, though perhaps `replaceOwner()` is not really necessary since `setOwner()` should be able to parse the body of the GH to find who the `oldOwner` was (if one existed).

mjasikowski replied:
Perhaps I could simplify the parameter description, but the idea is simple:

`newOwner` and `oldOwner` have values = we're replacing `oldOwner` with `newOwner`
`newOwner` is null, `OldOwner` has value  = we're deleting owner info from the issue
`newOwner` has value, `OldOwner` is null = we're adding a new owner to the issue
I can change the JSDoc description to:

```js
/**
 * Sets the owner of an issue
 * @param {String} newOwner to change to (if null, then the owner section will be removed from the issue)
 * @param {String} oldOwner to change from
 */
```
I couldn't get the old method (with scraping) to work reliably all the time, so I think this is a good middle ground (and also supports loading spinners, so you won't have to wonder if the browser registered your click fine).

tgolen replied:
OK, thanks for explaining that and I understand it better now. While logical, I still think that it's more complicated than it needs to be, and I prefer clarity over cleverness. I think it forces people to read the method doc in order to understand how to use the params, rather than letting the code be self-explanatory and not require any method doc explanations at all.

> I couldn't get the old method (with scraping) to work reliably all the time

It would be good to understand this better, because this new code is not doing anything different than the old code was. Maybe we're speaking of different things though. The GH body parsing and string replacement in this new method is not any different than the previous `setOnwer(), removeOner() and replaceOwner()` methods.

Adding the loading spinner is fine, I don't have any issues with that.

tgolen commented:
It looks like the K2 label buttons aren't working quite right.

Here, I've changed the issue from `Daily` to `Monthly` (but clicking on the "M" button) and it's left both labels on the issue.
![image](https://github.com/user-attachments/assets/a407aad3-8db5-4073-8f24-c67ee25d6099)

Here is a case where the "M" button isn't highlighted (like after refreshing the page).
![image](https://github.com/user-attachments/assets/c6039f02-b7ed-4d1d-b5e6-1c8cf3fbadec)

mjasikowski commented:
@tgolen reverted the previous assignee setting functions and fixed the K2 label selectors, would you mind giving a spin again?

neil-marcellini reviewed (changes requested):
One small fix please. Other than that it looks good assuming testing passes. Can you please write test steps even if it's just "load the K2 extension from this branch and test all features"?

mjasikowski commented on src/js/lib/pages/github/_base.js:
```diff
@@ -7,6 +8,54 @@ import $ from 'jquery';
+
+            if (!fileContents) {
+                console.error(`Could not load contents of ${checklistUrl} for some reason`);
+                return;
```
This was the original code moved to a different function, but I changed it to:

```js
        try {
            // Fetch the checklist contents
            const response = await fetch(checklistUrl);
            const fileContents = await response.text();

            // Call the API to add the comment
            await API.addComment(fileContents);
        } catch (error) {
            console.error('Error fetching the checklist:', error);
        } finally {
            // Restore the original button content
            button.innerHTML = originalContent;
        }
```
this way the loader will always be reverted, and:
1. If `fetch()` returns something other than 200, it will throw anyway
2. If fetched text is empty, then `API.addComment()` will throw due to incorrect parameters

So all error cases are covered now.

mjasikowski commented:
@neil-marcellini basic test steps added and all change requests were addressed

neil-marcellini reviewed (approved):
Good to go! Thanks for your hard work.
